src/client/conn.rs

```diff
@@ -0,0 +1,456 @@
+{
+    Builder::new()
+        .handshake(io)
+}
+
+/// The sender side of an established connection.
+pub struct SendRequest<B> {
+    dispatch: dispatch::Sender<proto::dispatch::ClientMsg<B>, ::Response>,
+
+}
+
+/// A future that processes all HTTP state for the IO object.
+///
+/// In most cases, this should just be spawned into an executor, so that it
+/// can process incoming and outgoing messages, notice hangups, and the like.
+pub struct Connection<T, B>
+where
+    T: AsyncRead + AsyncWrite,
+    B: Stream<Error=::Error> + 'static,
+    B::Item: AsRef<[u8]>,
+{
+    inner: proto::dispatch::Dispatcher<
+        proto::dispatch::Client<B>,
+        B,
+        T,
+        B::Item,
+        proto::ClientUpgradeTransaction,
+    >,
+}
+
+
+/// A builder to configure an HTTP connection.
+///
+/// After setting options, the builder is used to create a `Handshake` future.
+#[derive(Clone, Debug)]
+pub struct Builder {
+    h1_writev: bool,
+    keep_alive: bool,
+}
+
+/// A future setting up HTTP over an IO object.
+///
+/// If successful, yields a `(SendRequest, Connection)` pair.
+pub struct Handshake<T, B> {
+    inner: HandshakeInner<T, B, proto::ClientUpgradeTransaction>,
+}
+
+/// A future returned by `SendRequest::send_request`.
+///
+/// Yields a `Response` if successful.
+pub struct ResponseFuture {
+    // for now, a Box is used to hide away the internal `B`
+    // that can be returned if canceled
+    inner: Box<Future<Item=Response, Error=::Error>>,
+}
+
+/// Deconstructed parts of a `Connection`.
+///
+/// This allows taking apart a `Connection` at a later time, in order to
+/// reclaim the IO object, and additional related pieces.
+#[derive(Debug)]
+pub struct Parts<T> {
+    /// The original IO object used in the handshake.
+    pub io: T,
+    /// A buffer of bytes that have been read but not processed as HTTP.
+    ///
+    /// For instance, if the `Connection` is used for an HTTP upgrade request,
+    /// it is possible the server sent back the first bytes of the new protocol
+    /// along with the response upgrade.
+    ///
+    /// You will want to check for any existing bytes if you plan to continue
+    /// communicating on the IO object.
+    pub read_buf: Bytes,
+    _inner: (),
+}
+
+// internal client api
+
+/*
+pub(super) type ConnectionNoUpgrades<T, B> = proto::dispatch::Dispatcher<
+    proto::dispatch::Client<B>,
+    B,
+    T,
+    B::Item,
+    proto::ClientUpgradeTransaction,
+>;
+*/
+
+pub(super) struct HandshakeNoUpgrades<T, B> {
+    inner: HandshakeInner<T, B, proto::ClientTransaction>,
+}
+
+struct HandshakeInner<T, B, R> {
+    builder: Builder,
+    io: Option<T>,
+    _marker: PhantomData<(B, R)>,
+}
+
+// ===== impl SendRequest
+
+impl<B> SendRequest<B>
+{
+    /// Polls to determine whether this sender can be used yet for a request.
+    ///
+    /// If the associated connection is closed, this returns an Error.
+    pub fn poll_ready(&mut self) -> Poll<(), ::Error> {
```

Comment:
Is the intent for this that you'd call it when pulling a connection out of a pool?

Reply:
That is a fine time to do so, or even if you're not using a pool, and just have a single connection, it lets you know if the connection could process it right now. I think I need to alter this PR slightly so that it returns `NotReady` for h1 connections that aren't pipelining. It would also return `NotReady` if this were an h2 connection and you'd reached `SETTINGS_MAX_CONCURRENT_STREAMS `.

Reply:
If you did try to send a request, would that just be queued up on the connection becoming ready, or would it be an error?

Reply:
If the connection weren't ready, you'd get an error. We start using this `poll_ready` -> `send` pattern in the h2 crate, and it felt very natural, and it's what `Sink` is changing to in futures 0.2.

Reply:
Ah cool, makes sense. I was thinking it wouldn't be an error, so I was a bit confused :)